import axios, { AxiosRequestConfig } from 'axios'

import ResponseError from './ResponseError'

const api = axios.create({
  baseURL: process.env.EXPO_PUBLIC_API_URL,
  withCredentials: true,
})

interface Params {
  data?: object
  config?: AxiosRequestConfig
  url: string
}

api.interceptors.request.use(
  async (request) => {
    console.log(
      `\n\nüìÑ | BASE URL\n${request.baseURL || ''}\n\nüìÑ | ENDPOINT\n${
        request.url
      }\n\nüìÑ | HEADERS\n${JSON.stringify(request.headers)}\n\nüìÑ | BODY\n${
        JSON.stringify(request.data) || ''
      }\n\n`,
    )

    return request
  },
  (error) => {
    return Promise.reject(error)
  },
)

export async function get<R>(data: Params) {
  try {
    const response = await api.get<R>(data.url, {
      ...data.config,
      headers: {
        ...data.config?.headers,
      },
    })

    return response.data
  } catch (error: unknown) {
    throw new ResponseError(error)
  }
}

export async function post({ url, data, config }: Params) {
  try {
    const response = await api.post(
      url,
      {
        ...data,
      },
      {
        ...config,
        headers: {
          ...config?.headers,
        },
      },
    )

    return response.data
  } catch (error: unknown) {
    if (axios.isAxiosError(error)) throw new ResponseError(error)
    throw new Error('N√£o foi poss√≠vel conectar ao servidor')
  }
}

export default api
